### PR TITLE
iox-#264 fix char array assignment to cxx::string

### LINKED ISSUE #264: cxx::string doesn't respect the null termination in an assigned char array
## Required information

**Operating system:**
all

**Compiler version:**
all

**Observed result or behaviour:**
the following code fails
```
char foo[50] = "bar";
cxx::string<100> bar(foo);
assert(bar.size() == strnlen(foo, 50));
```
`bar.size()` would actually return 50, the size of the char array.

**Expected result or behaviour:**
The code above doesn't fail.

**Conditions where it occurred / Performed steps:**
create a char array (not string literal) which has a null termination and assign it to a cxx::string.

### PR DESCRIPTION
**Pull request by elBoberido**
closes #264 

Signed-off-by: Kraus Mathias (CC-AD/ESW1) <mathias.kraus2@de.bosch.com>

### CODE REVIEW

File: iceoryx_utils/include/iceoryx_utils/internal/cxx/string.inl [hunk 1]
```diff
@@ -119,9 +119,17 @@ inline string<Capacity>& string<Capacity>::operator=(const char (&rhs)[N]) noexc
...
-    m_rawstring[N - 1u] = '\0';
-    m_rawstringSize = N - 1u;
+
+    m_rawstringSize = strnlen(rhs, Capacity);
+    std::memcpy(m_rawstring, rhs, m_rawstringSize);
+    m_rawstring[m_rawstringSize] = '\0';
+
+    if (rhs[m_rawstringSize] != '\0')
+    {
+        std::cerr << "iox::cxx::string: Assignment of array which is not zero-terminated! Last value of array "
```
Comment (budrus) [inline iceoryx_utils/include/iceoryx_utils/internal/cxx/string.inl:129]: time for an error handling workshop. OK for now

**Review by budrus — APPROVED** [review]

File: iceoryx_utils/include/iceoryx_utils/internal/cxx/string.inl [hunk 2]
```diff
@@ -119,9 +119,17 @@ inline string<Capacity>& string<Capacity>::operator=(const char (&rhs)[N]) noexc
     {
         return *this;
     }
-    std::memcpy(m_rawstring, rhs, N - 1u);
-    m_rawstring[N - 1u] = '\0';
-    m_rawstringSize = N - 1u;
+
+    m_rawstringSize = strnlen(rhs, Capacity);
+    std::memcpy(m_rawstring, rhs, m_rawstringSize);
+    m_rawstring[m_rawstringSize] = '\0';
```
Comment (MatthiasKillat) [inline iceoryx_utils/include/iceoryx_utils/internal/cxx/string.inl:125]: Should be safe since m_rawstringSize is at most Capacity (due to strlen), so it is ok.

File: iceoryx_utils/test/moduletests/test_cxx_string.cpp [hunk 3]
```diff
@@ -361,6 +361,23 @@ TEST(String100, UnsafeCharToStringConstrIncludingNullCharWithCountResultsInSizeC
     EXPECT_THAT(testSubject.c_str(), StrEq("ice\0ryx"));
 }
 
+TEST(CharArrayAssignment, AssignZeroTerminatedCharArrayOfSizeForFullCapa)
```
Comment (MatthiasKillat) [inline iceoryx_utils/test/moduletests/test_cxx_string.cpp:364]: The test name confuses me, but ok. OfSize? SizeCapacity? How about AssignZeroTerminatedCharArrayOfSizeFullCapacity. Does not really matter though, those names are bound to be strange.

**Review by MatthiasKillat — APPROVED** [review]

File: iceoryx_utils/include/iceoryx_utils/internal/cxx/string.inl [hunk 4]
```diff
@@ -119,9 +119,17 @@ inline string<Capacity>& string<Capacity>::operator=(const char (&rhs)[N]) noexc
     {
```
Comment (marthtz) [inline iceoryx_utils/include/iceoryx_utils/internal/cxx/string.inl:119]: I know you just fixed a bug but wouldn't it be clearer with just a single return and write
```
if (c_str() != rhs)
{
  // add null termination
}
return * this;
```

File: iceoryx_utils/test/moduletests/test_cxx_string.cpp (hunk 3 above)
Comment (marthtz) [reply iceoryx_utils/test/moduletests/test_cxx_string.cpp:364]: Yeah, basically it's not telling what the size is, CharArrayOfSize???
I also prefer capacity.

File: iceoryx_utils/test/moduletests/test_cxx_string.cpp [hunk 5]
```diff
@@ -361,6 +361,23 @@ TEST(String100, UnsafeCharToStringConstrIncludingNullCharWithCountResultsInSizeC
...
+    string<7> testSubject(testString);
+    EXPECT_THAT(testSubject.size(), Eq(7));
+    EXPECT_THAT(testSubject.c_str(), StrEq("iceoryx"));
+}
+
+TEST(CharArrayAssignment, AssignNonZeroTerminatedCharArrayOfSizeForFullCapa)
+{
+    char testString[8] = "iceoryx";
+    testString[7] = 'x'; // overwrite the 0 termination
+    string<7> testSubject(testString);
```
Comment (marthtz) [inline iceoryx_utils/test/moduletests/test_cxx_string.cpp:376]: You could also just assign it to a smaller string?

File: iceoryx_utils/test/moduletests/test_cxx_string.cpp (hunk 3 above)
Comment (elBoberido) [reply iceoryx_utils/test/moduletests/test_cxx_string.cpp:372]: how about AssignZeroTerminatedCharArray**With**SizeForFullCapa?

File: iceoryx_utils/test/moduletests/test_cxx_string.cpp (hunk 5 above)
Comment (elBoberido) [reply iceoryx_utils/test/moduletests/test_cxx_string.cpp:384]: that's not possible. This is a corner case with the knowledge of the implementation. The c'tor accepts char array with a size of `Capacity + 1` at max

File: iceoryx_utils/include/iceoryx_utils/internal/cxx/string.inl (hunk 4 above)
Comment (elBoberido) [reply iceoryx_utils/include/iceoryx_utils/internal/cxx/string.inl:119]: probably, but I wanted to change as less as possible

**Review by dkroenke — APPROVED** [review]

File: iceoryx_utils/test/moduletests/test_cxx_string.cpp (hunk 5 above)
Comment (marthtz) [reply iceoryx_utils/test/moduletests/test_cxx_string.cpp:384]: Ok. Good to know!

File: iceoryx_utils/include/iceoryx_utils/internal/cxx/string.inl (hunk 4 above)
Comment (marthtz) [reply iceoryx_utils/include/iceoryx_utils/internal/cxx/string.inl:119]: Yeah. Maybe this comment fits better in PR #253 .

File: iceoryx_utils/test/moduletests/test_cxx_string.cpp (hunk 3 above)
Comment (marthtz) [reply iceoryx_utils/test/moduletests/test_cxx_string.cpp:372]: Better

Comment (MatthiasKillat) [reply iceoryx_utils/test/moduletests/test_cxx_string.cpp:372]: Fine with me.

**Review by MatthiasKillat — APPROVED** [review]

File: iceoryx_utils/test/moduletests/test_cxx_string.cpp (hunk 3 above)
Comment (elBoberido) [reply iceoryx_utils/test/moduletests/test_cxx_string.cpp:372]: maybe @FerdinandSpitzschnueffler has an even better idea. This could probably be renamed in #253
